tests/test_fix_course_delete_module_class_surgeon_test.php

```diff
@@ -0,0 +1,328 @@
+     * data exists.
+     * @covers ::reschedule_or_queue_adhoc_task
+     */
+    public function test_reschedule_or_queue_adhoc_task_different_data() {
+        $this->resetAfterTest(true);
+
+        $precount = count(\core\task\manager::get_adhoc_tasks('\core_course\task\course_delete_modules'));
+
+        // Schedule adhoc task.
+        $task = $this->removaltaskassign;
+        surgeon::reschedule_or_queue_adhoc_task($task);
+
+        // Schedule adhoc task for a different data.
+        $task = $this->removaltaskassign;
+        $quizdata = [
+            'cms' => [$this->quizcm],
+            'userid' => $this->user->id,
+            'realuserid' => $this->user->id
+        ];
+        $task->set_custom_data($quizdata);
+        surgeon::reschedule_or_queue_adhoc_task($task);
+
+        $this->assertEquals($precount + 2, count(\core\task\manager::get_adhoc_tasks('\core_course\task\course_delete_modules')));
+    }
+
+    /**
+     * Ensure that the reschedule_or_queue_adhoc_task function will not make any change for matching data if no time was
+     * specified.
+     * @covers ::reschedule_or_queue_adhoc_task
+     */
+    public function test_reschedule_or_queue_adhoc_task_match_no_change() {
+        $this->resetAfterTest(true);
+
+        $precount = count(\core\task\manager::get_adhoc_tasks('\core_course\task\course_delete_modules'));
+
+        // Schedule adhoc task.
+        $task = $this->removaltaskassign;
+        $task->set_next_run_time(time() + DAYSECS);
+        surgeon::reschedule_or_queue_adhoc_task($task);
+
+        $before = \core\task\manager::get_adhoc_tasks('\core_course\task\course_delete_modules');
+
+        // Schedule the task again but do not specify a time.
+        $task = $this->removaltaskassign;
+        surgeon::reschedule_or_queue_adhoc_task($task);
+
+        $this->assertEquals($precount + 1, count(\core\task\manager::get_adhoc_tasks('\core_course\task\course_delete_modules')));
+        $this->assertEquals($before, \core\task\manager::get_adhoc_tasks('\core_course\task\course_delete_modules'));
+    }
+
+    /**
+     * Ensure that the reschedule_or_queue_adhoc_task function will update the run time if there are planned changes.
+     * @covers ::reschedule_or_queue_adhoc_task
+     */
+    public function test_reschedule_or_queue_adhoc_task_match_update_runtime() {
+        $this->resetAfterTest(true);
+        $initialruntime = time() + DAYSECS;
+        $newruntime = time() + WEEKSECS;
+
+        // Schedule adhoc task.
+        $task = $this->removaltaskassign;
+        $task->set_next_run_time($initialruntime);
+        surgeon::reschedule_or_queue_adhoc_task($task);
+
+        $before = \core\task\manager::get_adhoc_tasks('\core_course\task\course_delete_modules');
+
+        // Schedule the task again.
+        $task = $this->removaltaskassign;
+        $task->set_next_run_time($newruntime);
+        surgeon::reschedule_or_queue_adhoc_task($task);
```

Comment:
It seems like these tests are valid for Moodle 3.5-3.6 only. `surgeon::reschedule_or_queue_adhoc_task()` is not supposed to be called in Moodle 3.7 onwards.
The best unit test here would be to fail the task, then fix it using public API and then assert that the task was rescheduled. Such unit test would work in differently in Moodle 3.5-3.6 (using `surgeon::reschedule_or_queue_adhoc_task()`) vs Moodle 3.7 onwards (`using cron API`), but the outcome will be the same and that's exactly what we need to test here.

Reply:
the function is refactored to call the `\core\task\manager::reschedule_or_queue_adhoc_task()` if it exists, otherwise very similar code will be called from within the surgeon class.
phpunittests confirm that tasks are rescheduled for all Moodle versions from 3.5+